pymtl3/stdlib/rtl/queues.py

```diff
@@ -156,3 +444,170 @@ def construct( s, MsgType, num_entries=2 ):
+
+    @s.update_on_edge
+    def up_full():
+      if s.reset:
+        s.full = b1(0)
+      elif ~s.full and s.enq.en:
+        s.full = b1(1)
+      elif s.full and s.deq.en:
+        s.full = b1(0)
+      else:
+        s.full = s.full
+    
+    # TODO: figure out whether to use deepcopy here.
+    @s.update_on_edge
+    def up_entry():
+      s.entry = s.enq.msg if s.enq.en else s.entry
+
+    @s.update
+    def up_enq_rdy():
+      if ~s.reset:
+        s.enq.rdy = b1(1) if ~s.full else b1(0)
+      else:
+        s.enq.rdy = b1(0)
+
+    @s.update
+    def up_deq_rdy():
+      if ~s.reset:
+        s.deq.rdy = b1(1) if s.full else b1(0)
+      else:
+        s.deq.rdy = b1(0)
+
+    s.connect( s.entry, s.deq.msg )
+
+  def line_trace( s ):
+    return "{}({}){}".format( s.enq, s.full, s.deq )
+
+#-------------------------------------------------------------------------
+# PipeQueue1EntryRTL
+#-------------------------------------------------------------------------
+
+class PipeQueue1EntryRTL( Component ):
+
+  def construct( s, EntryType ):
+
+    # Interface
+
+    s.enq = EnqIfcRTL( EntryType )
+    s.deq = DeqIfcRTL( EntryType )
+
+    # Components
+
+    s.entry = Wire( EntryType )
+    s.full  = Wire( Bits1 )
+
+    # Logic
+
+    @s.update_on_edge
+    def up_full():
+      if s.reset:
+        s.full = b1(0)
+      elif ~s.full and s.enq.en and ~s.deq.en:
+        s.full = b1(1)
+      elif s.full and ~s.enq.en and s.deq.en:
+        s.full = b1(0)
+      else:
+        s.full = s.full
+    
+    # TODO: figure out whether to use deepcopy here.
+    @s.update_on_edge
+    def up_entry():
+      s.entry = s.enq.msg if s.enq.en else s.entry
+
+    @s.update
+    def up_enq_rdy():
+      if ~s.reset:
+        s.enq.rdy = b1(1) if ~s.full or s.deq.en else b1(0)
+      else:
+        s.enq.rdy = b1(0)
+
+    @s.update
+    def up_deq_rdy():
+      if ~s.reset:
+        s.deq.rdy = b1(1) if s.full else b1(0)
+      else:
+        s.deq.rdy = b1(0)
+
+    s.connect( s.entry, s.deq.msg )
+
+  def line_trace( s ):
+    return "{}({}){}".format( s.enq, s.full, s.deq )
+
+#-------------------------------------------------------------------------
+# BypassQueue1EntryRTL
+#-------------------------------------------------------------------------
+
+class BypassQueue1EntryRTL( Component ):
+
+  def construct( s, EntryType ):
+
+    # Interface
+
+    s.enq = EnqIfcRTL( EntryType )
+    s.deq = DeqIfcRTL( EntryType )
+
+    # Components
+
+    s.entry = Wire( EntryType )
+    s.full  = Wire( Bits1 )
+
+    # Logic
+
+    @s.update_on_edge
+    def up_full():
+      if s.reset:
+        s.full = b1(0)
+      elif ~s.full and s.enq.en and ~s.deq.en:
+        s.full = b1(1)
+      elif s.full and ~s.enq.en and s.deq.en:
+        s.full = b1(0)
+      else:
+        s.full = s.full
+    
+    # TODO: figure out whether to use deepcopy here.
+    @s.update_on_edge
+    def up_entry():
+      s.entry = s.enq.msg if s.enq.en and ~s.deq.en else s.entry
```

Comment:
if we use deepcopy it's going to be ```if s.enq.en and ~s.deq.en: s.entry = deepcopy(s.enq.msg)```

Reply:
should we use deep copy here though? not sure

Reply:
I mean the general question you want to understand is, when do you need deepcopy? In this case, if s.enq.msg is a struct, when you assignit to s.entry without deepcopy, s.entry stores a pointer that points to the struct object. Let's say and this cycle your queue is full, so s.entry the struct object remain unchanged, and somehow some other update block in the producer unit change some field of this object (like s.imem.req.addr = pc+4). It will change the field of the object and unfortunately s.entry points to the same object if we don't do deepcopy. As a result s.entry is magically changed which is not what you want. This is one scenario. The other scenario is that s.entry points to a Bits object, and someone else do s.enq.msg[1:3] = 2. s.entry points to the same object without deepcopy.